astropy/io/ascii/tests/test_tdat.py

```diff
@@ -0,0 +1,817 @@
+
+
+def test_write_read_roundtrip():
+    """
+    Write a full-featured table with all types and see that it round-trips on
+    readback.
+    TDAT -> Table -> TDAT (formatted) -> Table
+    Tables should have the same content.
+    """
+    t = test_table
+    out = StringIO()
+    t.write(out, format="ascii.tdat")
+
+    t2s = [
+        Table.read(out.getvalue(), format="ascii.tdat"),
+        ascii.read(out.getvalue(), format="tdat"),
+    ]
+    for t2 in t2s:
+        assert t.meta == t2.meta
+        for name in t.colnames:
+            assert t[name].attrs_equal(t2[name])
+            assert np.all(t[name] == t2[name]) or np.all(t[name].mask == t2[name].mask)
+
+
+def test_write_read_roundtrip_empty_table(tmp_path):
+    """see https://github.com/astropy/astropy/issues/13191"""
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=TdatFormatWarning)
+        sfile = tmp_path / "x.tdat"
+        Table().write(sfile)
+        t = Table.read(sfile)
+        assert len(t) == 0
+        assert len(t.colnames) == 0
+
+
+def test_keyword_quotes():
+    lines = copy.copy(SIMPLE_LINES)
+    # double quotes
+    lines[1] = 'table_name = "astropy_table"'
+    t = Table.read(lines, format="ascii.tdat")
+    assert t.meta["keywords"]["table_name"] == "astropy_table"
+    # single quotes
+    lines[1] = "table_name = 'astropy_table'"
+    t = Table.read(lines, format="ascii.tdat")
+    assert t.meta["keywords"]["table_name"] == "astropy_table"
+    # back quotes
+    lines[1] = "table_name = `astropy_table`"
+    t = Table.read(lines, format="ascii.tdat")
+    assert t.meta["keywords"]["table_name"] == "astropy_table"
+    # combination and multiple, nested properly
+    lines[1] = "table_name = \"'`astropy_table`'\""
+    t = Table.read(lines, format="ascii.tdat")
+    assert t.meta["keywords"]["table_name"] == "astropy_table"
+
+    # mismatched
+    lines[1] = "table_name = \"astropy_table'"
+    with pytest.raises(TdatFormatError, match="Mismatched"):
+        t = Table.read(lines, format="ascii.tdat")
+    # combination, nested improperly
+    lines[1] = "table_name = \"'astropy_table\"'"
+    with pytest.raises(TdatFormatError, match="Mismatched"):
+        t = Table.read(lines, format="ascii.tdat")
+
+
+def test_tablenames():
+    """_summary_"""
+    lines = copy.copy(SIMPLE_LINES)
+    t = Table.read(lines, format="ascii.tdat")
+    out = StringIO()
+
+    with pytest.warns(TdatFormatWarning, match="'table_name' is too long"):
+        t.meta["table_name"] = "origin_12345678901234567890"
+        t.write(out, format="ascii.tdat")
+
+
+def test_bad_delimiter():
+    """
+    Passing a delimiter other than | (pipe) gives an exception
+    """
+    out = StringIO()
+
+    with pytest.warns(
+        TdatFormatWarning, match="Delimiters other than the pipe character"
+    ):
+        test_table.write(out, format="ascii.tdat", delimiter=",")
+
+    test_table.meta["keywords"]["field_delimiter"] = ","
```

Comment:
This modification of the table without making a copy caused problems with the double-test run. See #18025 for a fix. What I don't understand is how the tests could have passed before...